keadm/cmd/keadm/app/cmd/edge/upgrade.go

```diff
@@ -0,0 +1,397 @@
+	upgradev1alpha1 "github.com/kubeedge/kubeedge/pkg/apis/upgrade/v1alpha1"
+	"github.com/kubeedge/kubeedge/pkg/image"
+)
+
+// NewEdgeUpgrade returns KubeEdge edge upgrade command.
+func NewEdgeUpgrade() *cobra.Command {
+	upgradeOptions := newUpgradeOptions()
+
+	cmd := &cobra.Command{
+		Use:   "upgrade",
+		Short: "Upgrade edge component. Upgrade the edge node to the desired version.",
+		Long:  "Upgrade edge component. Upgrade the edge node to the desired version.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// upgrade edgecore
+			return upgradeOptions.upgrade()
+		},
+	}
+
+	addUpgradeFlags(cmd, upgradeOptions)
+	return cmd
+}
+
+// newJoinOptions returns a struct ready for being used for creating cmd join flags.
+func newUpgradeOptions() *UpgradeOptions {
+	opts := &UpgradeOptions{}
+	opts.ToVersion = "v" + common.DefaultKubeEdgeVersion
+	opts.Config = constants.DefaultConfigDir + "edgecore.yaml"
+
+	return opts
+}
+
+func (up *UpgradeOptions) upgrade() error {
+	// get EdgeCore configuration from edgecore.yaml config file
+	data, err := os.ReadFile(up.Config)
+	if err != nil {
+		return fmt.Errorf("failed to read config file %s: %v", up.Config, err)
+	}
+
+	configure := &v1alpha1.EdgeCoreConfig{}
+	err = yaml.Unmarshal(data, configure)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal config file %s: %v", up.Config, err)
+	}
+
+	upgrade := Upgrade{
+		UpgradeID:      up.UpgradeID,
+		HistoryID:      up.HistoryID,
+		FromVersion:    up.FromVersion,
+		ToVersion:      up.ToVersion,
+		Repository:     up.Repository,
+		ConfigFilePath: up.Config,
+		EdgeCoreConfig: configure,
+	}
+
+	defer func() {
+		// report upgrade result to cloudhub
+		if err := upgrade.reportUpgradeResult(); err != nil {
+			klog.Errorf("failed to report upgrade result to cloud: %v", err)
+		}
+		// cleanup idempotency record
+		if err := os.Remove(filepath.Join(util.KubeEdgePath, "idempotency_record")); err != nil {
+			klog.Errorf("failed to remove idempotency_record file: %v", err)
+		}
+	}()
+
+	// only allow upgrade when last upgrade finished
+	if util.FileExists(filepath.Join(util.KubeEdgePath, "idempotency_record")) {
+		upgrade.UpdateStatus(string(upgradev1alpha1.UpgradeFailedRollbackSuccess))
+		upgrade.UpdateFailureReason("last upgrade not finished, not allowed upgrade again")
+		return fmt.Errorf("last upgrade not finished, not allowed upgrade again")
+	}
+
+	// create idempotency_record file
+	os.MkdirAll(util.KubeEdgePath, 0755)
+	os.Create(filepath.Join(util.KubeEdgePath, "idempotency_record"))
+
+	// run script to do upgrade operation
+	err = upgrade.PreProcess()
+	if err != nil {
+		upgrade.UpdateStatus(string(upgradev1alpha1.UpgradeFailedRollbackSuccess))
+		upgrade.UpdateFailureReason(fmt.Sprintf("upgrade error: %v", err))
+		return fmt.Errorf("upgrade pre process failed: %v", err)
+	}
+
+	err = upgrade.Process()
+	if err != nil {
+		time.Sleep(1000 * time.Second)
```

Comment:
why need sleep 1000 second? it is not reliable

Reply:
will remove